libraries/lib-snapping/SnapUtils.cpp

```diff
@@ -0,0 +1,315 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  SnapUtils.cpp
+
+  Dmitry Vedenko
+
+**********************************************************************/
+
+#include "SnapUtils.h"
+
+#include <algorithm>
+#include <cmath>
+
+namespace
+{
+const wxString SnapModeKey = L"/Snap/Mode";
+const wxString SnapToKey = L"/Snap/To";
+const wxString OldSnapToKey = L"/SnapTo";
+const auto PathStart = L"SnapFunctions";
+
+struct RegistryVisitor : public Registry::Visitor
+{
+   explicit RegistryVisitor(SnapRegistryVisitor& _visitor)
+       : visitor(_visitor)
+   {
+   }
+   
+   void BeginGroup(Registry::GroupItem& item, const Path&) override
+   {
+      auto group = dynamic_cast<SnapRegistryGroup*>(&item);
+
+      if (group != nullptr)
+         visitor.BeginGroup(*group);
+   }
+
+   void EndGroup(Registry::GroupItem& item, const Path&) override
+   {
+      auto group = dynamic_cast<SnapRegistryGroup*>(&item);
+
+      if (group != nullptr)
+         visitor.EndGroup(*group);
+   }
+
+   void Visit(Registry::SingleItem& item, const Path&) override
+   {
+      auto concreteItem = dynamic_cast<SnapRegistryItem*>(&item);
+
+      if (concreteItem != nullptr)
+         visitor.Visit(*concreteItem);
+   }
+
+   SnapRegistryVisitor& visitor;
+};
+} // namespace
+
+StringSetting SnapToSetting { SnapToKey, "seconds" };
+
+EnumSetting<SnapMode> SnapModeSetting {
+   SnapModeKey,
+   EnumValueSymbols { L"OFF", L"NEAREST", L"PRIOR" },
+   0,
+   { SnapMode::SNAP_OFF, SnapMode::SNAP_NEAREST, SnapMode::SNAP_PRIOR }
+};
+
+SnapMode ReadSnapMode()
+{
+   if (gPrefs->HasEntry(SnapModeKey))
+      return SnapModeSetting.ReadEnum();
+
+   return static_cast<SnapMode>(gPrefs->Read(OldSnapToKey, 0L));
+}
+
+Identifier ReadSnapTo()
+{
+   if (gPrefs->HasEntry(SnapToKey))
+      return SnapToSetting.Read();
+
+   return {};
```

Comment:
Should that instead be
`return SnapToSetting.GetDefault();`

Reply:
That's unfortunate omission that I forgot to implement. This function will be quite a bit more complex.